### ISSUE TITLE
[fn][compute] Add support for Duration data type

### ISSUE THREAD

**Issue by aymkhalil** [datastax/pulsar-transformations#28]
Support ISO 8601 Duration type: https://en.m.wikipedia.org/wiki/ISO_8601#Durations

In Avro, this is supported as built in logical type: https://avro.apache.org/docs/1.10.2/spec.html#Duration

**Comment by cbornet** [comment]
> Drafting duration logical type: https://github.com/datastax/pulsar-transformations/commit/6b432173dceef10d9520186b13cd98c95b854434

With that we can add a field which is an AVRO duration to an AVRO datetime, is that correct ?

**Comment by cbornet** [comment]
> java.util.Duration accepts the ISO format PnYnMnD (notice no hours, min, seconds, etc)

Do you mean `java.util.time.Period` ?

**Comment by cbornet** [comment]
In the JDK, a Duration is an exact amount of time. So it can't be expressed in months or years because these have variable time lengths.
For this reason it's not possible to map an AVRO duration to a j.u.t.Duration.
A Period is a variable amount of time that takes into account leap days, Daylight-Saving-Time and so on. It would be more appropriate to map an AVRO duration LT. I don't know why Period doesn't have h/m/s.
Instead of using joda which is in maintenance mode, we could probably use [Threeten-extra lib's PeriodDuration](https://github.com/ThreeTen/threeten-extra/blob/master/src/main/java/org/threeten/extra/PeriodDuration.java) ?
Note that the diff between Duration and Period only appears on ZonedDateTime which we don't support atm so there shouldn't be much trouble.

**Comment by aymkhalil** [comment]
Yes in general we'll need a Period based implementation, will take a look at the lib you proposed.

**Comment by aymkhalil** [comment]
> > Drafting duration logical type: [6b43217](https://github.com/datastax/pulsar-transformations/commit/6b432173dceef10d9520186b13cd98c95b854434)
> 
> With that we can add a field which is an AVRO duration to an AVRO datetime, is that correct ?

The draft PR only proposes a new type called DURATION and maps to AVRO's logical duration type, the date utility methods for adding will need to be supported (the [plus](https://docs.oracle.com/javase/8/docs/api/java/time/LocalDateTime.html#plus-long-java.time.temporal.TemporalUnit-) for example which the dateadd methods relies on understands absolute unit in time but not periods)

**Comment by aymkhalil** [comment]
I'm starting to realize that the Avro logical type called Duration is actually a Period in java terms...

**Comment by cbornet** [comment]
Yes it is.
